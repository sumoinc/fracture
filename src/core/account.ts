import { Component } from "projen";
import { Organization } from "./organization";
import { OrganizationalUnit } from "./organizational-unit";
import { Region, RegionOptions } from "./region";

export interface AccountOptions {
  id: string;
  name?: string;
  rootEmail?: string;
  isManagementAccount?: boolean;
  organizationalUnit?: OrganizationalUnit;
}

export class Account extends Component {
  // member components
  public readonly regions: Region[];
  // parent
  public readonly organization: Organization;
  // all other options
  public readonly options: AccountOptions;

  constructor(organization: Organization, options: AccountOptions) {
    super(organization.project);

    /***************************************************************************
     *
     * INIT OPERATION
     *
     **************************************************************************/

    // member components
    this.regions = [];

    // parents + inverse
    this.organization = organization;
    this.organization.accounts.push(this);

    // all other options
    const defaultOptions: Partial<AccountOptions> = {
      isManagementAccount: false,
    };
    this.options = { ...defaultOptions, ...options };

    // if OU defined, add account to OU
    if (this.options.organizationalUnit) {
      this.options.organizationalUnit.accounts.push(this);
    }

    // debugging output
    this.project.logger.info(`INIT Account: "${this.id}"`);
  }

  public get id() {
    return this.options.id;
  }

  public get name() {
    return this.options.name;
  }

  public get rootEmail() {
    return this.options.rootEmail;
  }

  public get isManagementAccount() {
    return this.options.isManagementAccount;
  }

  public get organizationalUnit() {
    return this.options.organizationalUnit;
  }

  /*****************************************************************************
   *
   *  Configuration Helpers
   *
   ****************************************************************************/

  /**
   * Add an region to an account.
   *
   * @param {RegionOptions}
   * @returns {Region}
   */
  public addRegion(options: RegionOptions) {
    const region = this.regions.find((r) => r.id === options.id);
    return region ? region : new Region(this, options);
  }
}
