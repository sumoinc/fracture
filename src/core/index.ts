export * from "./access-pattern";
export * from "./account";
export * from "./audit-strategy";
export * from "./fracture";
export * from "./index";
export * from "./naming-strategy";
export * from "./operation";
export * from "./organization";
export * from "./organizational-unit";
export * from "./region";
export * from "./resource-attribute";
export * from "./resource";
export * from "./service";
export * from "./structure";
