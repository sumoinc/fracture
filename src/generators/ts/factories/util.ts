import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
