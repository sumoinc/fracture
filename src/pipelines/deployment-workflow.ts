import { Component } from "projen";
import { GitHub, GithubWorkflow } from "projen/lib/github";
import { Job, JobPermission, JobStep } from "projen/lib/github/workflows-model";
import { NodeProject } from "projen/lib/javascript";

const BUILD_JOBID = "build";

export interface DeploymentWorkflowOptions {
  /**
   * Name of the file (e.g. "deploy" becomes "deploy.yml").
   *
   * @default "deploy"
   */
  name?: string;

  workflowNodeVersion?: string;
  pnpmVersion?: string;
  mutableBuild?: boolean;
}

export class DeploymentWorkflow extends Component {
  /**
   * Returns a environment by name, or undefined if it doesn't exist
   */
  public static of(project: NodeProject): DeploymentWorkflow | undefined {
    const isDefined = (c: Component): c is DeploymentWorkflow =>
      c instanceof DeploymentWorkflow;
    return project.components.find(isDefined);
  }
  /**
   * Name of the file (e.g. "deploy" becomes "deploy.yml").
   *
   * @default "deploy"
   */
  readonly name: string;
  /**
   * Deployment Workflow
   */
  public readonly workflow: GithubWorkflow;
  private readonly github: GitHub;
  private readonly defaultRunners: string[] = ["ubuntu-latest"];
  private readonly workflowNodeVersion: string;
  private readonly pnpmVersion: string;
  private readonly mutableBuild: boolean;

  constructor(project: NodeProject, options: DeploymentWorkflowOptions) {
    super(project);

    const github = GitHub.of(project);

    if (!github) {
      throw new Error(
        "BuildWorkflow is currently only supported for GitHub projects"
      );
    }

    this.name = options.name ?? "deploy";
    this.workflow = new GithubWorkflow(github, this.name);
    this.github = github;
    this.workflowNodeVersion = options.workflowNodeVersion ?? "18";
    this.pnpmVersion =
      options.pnpmVersion ?? project.package.pnpmVersion ?? "18";
    this.mutableBuild = options.mutableBuild ?? false;

    if (this.github) {
      console.log("github is defined");
    }

    // build job
    //this.addBuildJob();

    //this.workflow.addJob(BUILD_JOBID, {})};
  }

  private renderBuildJob(): Job {
    const project = this.project as NodeProject;
    return {
      runsOn: this.defaultRunners,
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.WRITE,
      },
      steps: [
        ...this.renderBuildSetup(),
        {
          name: project.buildTask.name,
          run: this.github.project.runTaskCommand(project.buildTask),
        },
      ],
    };
  }

  private renderBuildSetup(): Array<JobStep> {
    const project = this.project as NodeProject;
    return [
      {
        name: "Checkout",
        uses: "actions/checkout@v3",
        with: {
          "fetch-depth": 0,
        },
      },
      {
        name: "Setup pnpm",
        uses: "pnpm/action-setup@v2.2.4",
        with: { version: this.pnpmVersion },
      },
      {
        name: "Setup Node.js",
        uses: "actions/setup-node@v3",
        with: {
          "node-version": this.workflowNodeVersion,
          cache: "pnpm",
        },
      },
      {
        name: "Install dependencies",
        run: this.mutableBuild
          ? project.package.installAndUpdateLockfileCommand
          : project.package.installCommand,
      },
    ];
  }

  preSynthesize(): void {
    this.workflow.addJob(BUILD_JOBID, this.renderBuildJob());
  }
}
