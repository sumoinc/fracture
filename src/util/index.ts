export * from "./determineBuildId";
export * from "./test-util";
