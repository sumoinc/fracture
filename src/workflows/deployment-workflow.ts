import { Component } from "projen";
import { NodeProject } from "projen/lib/javascript";
import { Workflow } from "./workflow";

export class DeploymentWorkflow extends Workflow {
  /**
   * Returns the deployment workflow for a project or creates one if it
   * doesn't exist yet. Singleton?
   */
  public static of(project: NodeProject): DeploymentWorkflow {
    const isDefined = (c: Component): c is DeploymentWorkflow =>
      c instanceof DeploymentWorkflow;
    return (
      project.components.find(isDefined) ?? new DeploymentWorkflow(project)
    );
  }

  constructor(public readonly project: NodeProject) {
    super(project, { name: "deployment" });
  }
}
