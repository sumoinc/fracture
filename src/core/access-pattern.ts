import { deepMerge } from "projen/lib/util";
import { ValueOf } from "type-fest";
import { Operation, OPERATION_SUB_TYPE } from "./operation";
import { Resource } from "./resource";
import {
  ResourceAttribute,
  ResourceAttributeGenerator,
  ResourceAttributeOptions,
} from "./resource-attribute";
import { FractureComponent } from "../core/component";
import { DynamoGsi } from "../dynamodb/dynamo-gsi";
import { DynamoTable } from "../dynamodb/dynamo-table";

export const ACCESS_PATTERN_TYPE = {
  /**
   * Main PK and PK pattern used to identify this record.
   */
  IDENTIFIER: "Identifier",
  /**
   * Lookup pattern for quick begins_with() searching and such.
   */
  LOOKUP: "lookup",
  /**
   * Any other AP
   */
  OTHER: "other",
} as const;

export interface AccessPatternOptions {
  dynamoGsi?: DynamoGsi;
  pkAttributeOptions?: ResourceAttributeOptions;
  skAttributeOptions?: ResourceAttributeOptions;
  type?: ValueOf<typeof ACCESS_PATTERN_TYPE>;
}

export class AccessPattern extends FractureComponent {
  // member components
  public pkAttribute: ResourceAttribute;
  public skAttribute: ResourceAttribute;
  public operations: Operation[] = [];
  // parent
  public readonly resource: Resource;
  // all other options
  public readonly options: Required<AccessPatternOptions>;

  constructor(resource: Resource, options: AccessPatternOptions) {
    super(resource.fracture);

    /***************************************************************************
     *
     * DEFAULT OPTIONS
     *
     **************************************************************************/

    // make a new gsi is one wasn't passed into the Access Pattern
    const dynamoGsi = options.dynamoGsi
      ? options.dynamoGsi
      : new DynamoGsi(resource.dynamoTable);

    const defaultOptions: Required<AccessPatternOptions> = {
      dynamoGsi,
      pkAttributeOptions: {
        name: dynamoGsi.pkName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      skAttributeOptions: {
        name: dynamoGsi.skName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      type: ACCESS_PATTERN_TYPE.OTHER,
    };

    /***************************************************************************
     *
     * INIT ACCESS PATTERN
     *
     **************************************************************************/

    // all other options
    this.options = deepMerge([
      defaultOptions,
      options,
    ]) as Required<AccessPatternOptions>;
    // the attribute might already be defined.
    const pkAttribute = resource.getAttributeByName(dynamoGsi.pkName);
    const skAttribute = resource.getAttributeByName(dynamoGsi.skName);

    // member components
    this.pkAttribute = pkAttribute
      ? pkAttribute
      : new ResourceAttribute(resource, this.options.pkAttributeOptions);
    this.skAttribute = skAttribute
      ? skAttribute
      : new ResourceAttribute(resource, this.options.skAttributeOptions);

    // parent + inverse
    this.resource = resource;
    this.resource.accessPatterns.push(this);

    this.project.logger.info(
      `Access Pattern for: "${this.pkAttribute.name}:${this.skAttribute.name}" initialized.`
    );

    return this;
  }

  public build() {
    this.project.logger.debug(
      `BUILD Access Pattern: "${this.pkAttribute.name}:${this.skAttribute.name}" called.`
    );
    /***************************************************************************
     *
     * CRUD OPERATIONS
     *
     * TODO
     *
     **************************************************************************/
    /*
    new Operation(this, {
      operationType: OPERATION_TYPE.MUTATION,
      operationSubType: OPERATION_SUB_TYPE.LIST,
    });
    */
  }

  addPkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.pkAttribute.compositionSources.push(sourceAttribute);
  }

  addSkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.skAttribute.compositionSources.push(sourceAttribute);
  }

  /**
   * Returns operation for given sub-type
   */
  public getOperation(
    operationSubType: ValueOf<typeof OPERATION_SUB_TYPE>
  ): Operation {
    const returnOperation = this.operations.find(
      (o) => o.operationSubType === operationSubType
    );

    if (!returnOperation) {
      throw new Error(`Operation not found for sub-type: ${operationSubType}`);
    }

    return returnOperation;
  }

  public get createOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.CREATE_ONE);
  }

  public get readOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.READ_ONE);
  }

  public get updateOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.UPDATE_ONE);
  }

  public get deleteOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.DELETE_ONE);
  }

  public get isKeyAccessPattern(): boolean {
    return this.dynamoTable.keyDynamoGsi === this.dynamoGsi;
  }

  public get isLookupAccessPattern(): boolean {
    return this.dynamoTable.lookupDynamoGsi === this.dynamoGsi;
  }

  public get dynamoTable(): DynamoTable {
    return this.dynamoGsi.dynamoTable;
  }

  public get dynamoGsi(): DynamoGsi {
    return this.options.dynamoGsi;
  }
}
