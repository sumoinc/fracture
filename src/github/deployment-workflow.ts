import { Component } from "projen";
import { GitHub, GithubWorkflow } from "projen/lib/github";
import { JobPermission, JobStep } from "projen/lib/github/workflows-model";
import { Fracture, Pipeline } from "../core";

const PULL_REQUEST_REF = "${{ github.event.pull_request.head.ref }}";
const PULL_REQUEST_REPOSITORY =
  "${{ github.event.pull_request.head.repo.full_name }}";
const BUILD_JOBID = "build";
//const SELF_MUTATION_STEP = "self_mutation";
//const SELF_MUTATION_HAPPENED_OUTPUT = "self_mutation_happened";

export interface DeploymentWorkflowOptions {
  /**
   * The name of the workflow.
   */
  name: string;
}

export class DeploymentWorkflow extends Component {
  private readonly github: GitHub;
  private readonly githubWorkflow: GithubWorkflow;
  private readonly options: Required<DeploymentWorkflowOptions>;

  constructor(pipeline: Pipeline, options: DeploymentWorkflowOptions) {
    super(pipeline.project);

    const github = GitHub.of(pipeline.project);
    if (!github) {
      throw new Error(
        "DeploymentWorkflow is currently only supported for GitHub projects"
      );
    }
    this.github = github;

    this.options = options;

    this.githubWorkflow = new GithubWorkflow(github, this.name);
    this.githubWorkflow.on({
      pullRequest: {},
      workflowDispatch: {}, // allow manual triggering
    });

    this.addBuildJob();
  }

  public get name() {
    return this.options.name;
  }

  private addBuildJob() {
    this.githubWorkflow.addJob(BUILD_JOBID, {
      runsOn: ["ubuntu-latest"],
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.WRITE,
      },
      steps: (() => this.renderBuildSteps()) as any,
      /*outputs: {
        [SELF_MUTATION_HAPPENED_OUTPUT]: {
          stepId: SELF_MUTATION_STEP,
          outputName: SELF_MUTATION_HAPPENED_OUTPUT,
        },
      },
      */
    });
  }

  /**
   * Called (lazily) during synth to render the build job steps.
   */
  private renderBuildSteps(): JobStep[] {
    const fracture = this.project as Fracture;
    const { buildTask, synthTask } = fracture.turborepo;
    //const buildTask = TurboRepo.buildTask(fracture);
    //const synthTask = TurboRepo.synthTask(fracture);
    return [
      {
        name: "Checkout",
        uses: "actions/checkout@v3",
        with: {
          ref: PULL_REQUEST_REF,
          repository: PULL_REQUEST_REPOSITORY,
          ...(this.github.downloadLfs ? { lfs: true } : {}),
        },
      },

      //...this.preBuildSteps,
      ...fracture.renderWorkflowSetup({
        mutable: true,
      }),

      {
        name: buildTask.name,
        run: this.github.project.runTaskCommand(buildTask),
      },

      {
        name: synthTask.name,
        run: this.github.project.runTaskCommand(synthTask),
        env: { foo: "bar" },
      },

      //...this.postBuildSteps,

      // check for mutations and upload a git patch file as an artifact
      /*
      ...WorkflowActions.uploadGitPatch({
        stepId: SELF_MUTATION_STEP,
        outputName: SELF_MUTATION_HAPPENED_OUTPUT,
        mutationError:
          "Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch.",
      }),
      */

      // upload the build artifact only if we have post-build jobs (otherwise, there's no point)
      /*
      ...(this._postBuildJobs.length == 0
        ? []
        : [
            {
              name: "Backup artifact permissions",
              continueOnError: true,
              run: `cd ${this.artifactsDirectory} && getfacl -R . > ${PERMISSION_BACKUP_FILE}`,
            },
            {
              name: "Upload artifact",
              uses: "actions/upload-artifact@v3",
              with: {
                name: BUILD_ARTIFACT_NAME,
                path: this.artifactsDirectory,
              },
            },
          ]),
          */
    ];
  }
}
