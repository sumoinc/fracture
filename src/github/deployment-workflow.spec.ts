import { LogLevel } from "projen";
import { synthSnapshot } from "projen/lib/util/synth";
import { DeploymentWorkflow } from "./deployment-workflow";
import { Fracture, REGION_IDENTITIER } from "../core";

/*******************************************************************************
 * TEST SETUP
 ******************************************************************************/

let workflow: DeploymentWorkflow;

beforeEach(() => {
  fracture = new Fracture({
    name: "test-project",
    logging: {
      level: LogLevel.OFF,
    },
  });
  const org = fracture.addOrganization({ id: "test-org" });
  const account = org.addAccount({ id: "id", name: "test-account" });
  const devUsEast = fracture.addEnvironment({
    account,
    region: REGION_IDENTITIER.US_EAST_1,
  });
  const devUsWest = fracture.addEnvironment({
    account,
    region: REGION_IDENTITIER.US_WEST_2,
  });
  const app = fracture.addApp({ name: "identity-service" });
  const pipeline = app.addPipeline({
    name: "feature-branch",
    branchTriggerPattern: "feature/*",
  });
  pipeline.addStage({ name: "East Coast", environment: devUsEast });
  pipeline.addStage({ name: "West Coast", environment: devUsWest });
  workflow = new DeploymentWorkflow(pipeline, { name: "deploy-test" });
});

/*******************************************************************************
 * DEPLOY WORKFLOW
 ******************************************************************************/

describe("workflow file tests", () => {
  test("Generates expected pipeline", () => {
    const workflow = synthWorkflow("deploy-test");
    expect(workflow).toMatchSnapshot();

    console.log(workflow);
    /* const fileContent =
      workflows[`.github/workflows/${pipeline.deployName}.yml`];
    console.log(workflows[`.github/workflows/${pipeline.deployName}.yml`]);
    expect(fileContent).toBeDefined();
    expect(fileContent).toMatchSnapshot();
    */
  });
});

describe("formatting", () => {
  test("Deployments depend on build", () => {});
  test("Deployments depend on branch", () => {});
});

/**
 *
 * Helper function to generate a workflow file so we can test it's structure.
 *
 */

function synthWorkflow(name: string): any {
  const snapshot = synthSnapshot(workflow.project);
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(`.github/workflows/${workflow.name}`))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {});
  return filtered;
}
