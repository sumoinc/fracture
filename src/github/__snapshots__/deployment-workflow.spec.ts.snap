// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default workflow no args turborepo 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".

name: deployment
on:
  push:
    branches:
      - main
    paths: []
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v2.2.4
        with:
          version: \\"8\\"
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: \\"18\\"
          cache: pnpm
      - name: Install dependencies
        run: pnpm i --no-frozen-lockfile
      - name: Build
        run: npx projen turbo:build
      - name: Synth
        run: npx projen turbo:synth
  deploy:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Deploy
        run: npx projen turbo:deploy
"
`;

exports[`Default workflow no args turborepo 2`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".

name: deployment
on:
  push:
    branches:
      - main
    paths: []
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v2.2.4
        with:
          version: \\"8\\"
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: \\"18\\"
          cache: pnpm
      - name: Install dependencies
        run: pnpm i --no-frozen-lockfile
      - name: Build
        run: npx projen turbo:build
      - name: Synth
        run: npx projen turbo:synth
  deploy:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Deploy
        run: npx projen turbo:deploy
"
`;

exports[`Default workflow with TurboRepo 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".

name: with-turbo-repo
on:
  push:
    branches:
      - main
    paths: []
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v2.2.4
        with:
          version: \\"8\\"
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: \\"18\\"
          cache: pnpm
      - name: Install dependencies
        run: pnpm i --no-frozen-lockfile
      - name: Build
        run: npx projen build
      - name: Synth
        run: npx projen synth
  deploy:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Deploy
        run: npx projen deploy
"
`;

exports[`Default workflow without TurboRepo 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".

name: without-turbo-repo
on:
  push:
    branches:
      - main
    paths: []
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v2.2.4
        with:
          version: \\"8\\"
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: \\"18\\"
          cache: pnpm
      - name: Install dependencies
        run: pnpm i --no-frozen-lockfile
      - name: Build
        run: npx projen turbo:build
      - name: Synth
        run: npx projen turbo:synth
  deploy:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Deploy
        run: npx projen turbo:deploy
"
`;

exports[`Waves & Stages one wave, one stage 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".

name: deployment
on:
  push:
    branches:
      - main
    paths: []
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v2.2.4
        with:
          version: \\"8\\"
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: \\"18\\"
          cache: pnpm
      - name: Install dependencies
        run: pnpm i --no-frozen-lockfile
      - name: Build
        run: npx projen turbo:build
      - name: Synth
        run: npx projen turbo:synth
  deploy:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    env:
      CI: \\"true\\"
    steps:
      - name: Deploy
        run: npx projen turbo:deploy
"
`;
